### PR TITLE
feat: local binary update

### PR DESCRIPTION
**Pull request by maxwelbm**
- update binary local
- update suggestion

### CODE REVIEW

**Comment by maxwelbm** [comment]
This is a Draft

**Review by gabrielnes-azion — DISMISSED** [review]

File: pkg/upbin/upbin.go [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+	"github.com/aziontech/azion-cli/pkg/cmd/version"
+)
+
+type system string
+
+const (
+	azioncli string = "azioncli"
+
+	windows system = "windows"
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:32]: I don't think we currently build the binaries for windows... is this just for future proofing?

File: pkg/upbin/upbin.go [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+	darwin  system = "darwin"
+	linux   system = "linux"
+	unknown system = "unknown"
+
+	repoURL string = "https://github.com/aziontech/azion-cli.git"
+	url     string = "https://downloads.azion.com/%s/x86_64/azioncli"
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:38]: not every binary is x86_64, some of them are arm64 and/or arm32

File: pkg/upbin/upbin.go [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+	LastActivity    string = "LAST_ACTIVITY"
+	packageAzioncli string = "aziontech/tap/azioncli"
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:41]: rename this var to `tapAzioncli`, because this is only used for brew tap 😄

File: pkg/upbin/upbin.go [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+	return nil
+}
+
+func GetCurrentVersion() int {
+	n, _ := Format(version.BinVersion)
+	return n
+}
+
+func GetLatestVersion() int {
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:114]: Can't we create a function that both this and the one in init command use?

File: pkg/upbin/upbin.go [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+}
+
+func which(command string) (string, error) {
+	path := os.Getenv("PATH")
+	paths := filepath.SplitList(path)
+
+	for _, dir := range paths {
+		executablePath := filepath.Join(dir, command)
+		_, err := os.Stat(executablePath)
+		if err == nil {
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:179]: always check if err != nil

File: pkg/upbin/upbin.go [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+		return windows
+	default:
+		return unknown
+	}
+}
+
+func prepareURL() (string, error) {
+	sys := GetSystem()
+	if sys == unknown {
+		return "", errors.New("unknown system")
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:203]: maybe we can put this in an error message variable

File: pkg/upbin/upbin.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+	return fmt.Sprintf(url, sys), nil
+}
+
+func needToUpdate() bool {
+	return GetLatestVersion() >= GetCurrentVersion()
+}
+
+func wantToUpdate() bool {
+	prompt := promptui.Select{
+		Label: "Do you want to update 'azioncli'?",
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:214]: Maybe:
"A new version of 'azioncli' was published. Do you wish to update to the latest version?"
or something like that

File: pkg/upbin/upbin.go [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+			return executablePath, nil
+		}
+	}
+
+	return "", fmt.Errorf("command '%s' not found", command)
+}
+
+func GetSystem() system {
+	switch system(runtime.GOOS) {
+	case linux:
```
Comment (PatrickMenoti) [inline pkg/upbin/upbin.go:189]: You are checking this here, but you never used the information. 
I think your original idea was to get the OS, so then you can run the correct installation procedure for their OS
e.g.: if you verified they are using mac, you can try to run "brew install aziontech/tap/azioncli", then if it fails, you just replace the binary where it's installed. 
At least, that's what I remember from our previous refinements

File: pkg/upbin/upbin.go (hunk 1 above)
Comment (maxwelbm) [reply pkg/upbin/upbin.go:32]: we still don't support it but I already left the type created for future features

File: pkg/upbin/upbin.go (hunk 4 above)
Comment (maxwelbm) [reply pkg/upbin/upbin.go:114]: The idea is for init to use this function, I just didn't apply it 🪜

File: pkg/upbin/upbin.go (hunk 5 above)
Comment (maxwelbm) [reply pkg/upbin/upbin.go:179]: I agree 👍🏽

File: pkg/upbin/upbin.go (hunk 8 above)
Comment (maxwelbm) [reply pkg/upbin/upbin.go:189]: 🐴 🐴 🐴

File: pkg/upbin/packages_managers.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+func getFileNameFromURL(url string) string {
+	parts := strings.Split(url, "/")
+	return parts[len(parts)-1]
+}
+
+func installPackage(fileName string) error {
+	ext := getFileExtension(fileName)
+	switch ext {
```
Comment (PatrickMenoti) [inline pkg/upbin/packages_managers.go:139]: We also have 'apk'

**Review by PatrickMenoti — DISMISSED** [review]

**Review by ftoldato — DISMISSED** [review]

File: utils/errors.go [hunk 10]
```diff
@@ -39,4 +41,7 @@ var (
 	ErrorTimeoutAPICall         = errors.New("CLI's request has timed out during communication with Azion. Verify if it has completed successfully or wait some time and try the command again")
 	ErrorCreateFile             = errors.New("Failed to create %s file")
 	ErrorProductNotOwned        = errors.New("This account does not own the following product")
+	ErrorUnknownSystem          = errors.New("unknown system")
+	ErrorCommandNotFound        = errors.New("command '%s' not found")
```
Comment (gabriel-azion) [inline utils/errors.go:45]:
```suggestion
	ErrorCommandNotFound        = errors.New("Command '%s' not found")
```

File: utils/errors.go [hunk 11]
```diff
@@ -39,4 +41,7 @@ var (
...
+	ErrorGetAssetsNamesAzioncli = errors.New("failed to fetch the assets names from azioncli")
```
Comment (gabriel-azion) [inline utils/errors.go:46]:
```suggestion
	ErrorGetAssetsNamesAzioncli = errors.New("Failed to fetch the assets names from azioncli")
```

File: utils/errors.go [hunk 12]
```diff
@@ -39,4 +41,7 @@ var (
 	ErrorTimeoutAPICall         = errors.New("CLI's request has timed out during communication with Azion. Verify if it has completed successfully or wait some time and try the command again")
 	ErrorCreateFile             = errors.New("Failed to create %s file")
 	ErrorProductNotOwned        = errors.New("This account does not own the following product")
+	ErrorUnknownSystem          = errors.New("unknown system")
```
Comment (gabriel-azion) [inline utils/errors.go:44]:
```suggestion
	ErrorUnknownSystem          = errors.New("Unknown system")
```

File: utils/errors.go (hunk 10 above)
Comment (gabriel-azion) [inline utils/errors.go:45]:
```suggestion
	ErrorCommandNotFound        = errors.New("Command '%s' not found")
```

File: messages/root/messages.go [hunk 13]
```diff
@@ -7,6 +7,7 @@ azioncli [flags]
 	RootShortDescription = "Azion-CLI"
 	RootLongDescription  = "Calls the CLI via subcommands to interact with Azion services easily"
 	RootHelpFlag         = "Displays more information about the Azion CLI"
+	RootDoNotUpdate      = "not receive update notification"
```
Comment (gabriel-azion) [inline messages/root/messages.go:10]:
```suggestion
	RootDoNotUpdate      = "Do not receive update notification"
```

**Review by gabriel-azion — DISMISSED** [review]

**Review by gabriel-azion — APPROVED** [review]

**Review by pauloSF0 — APPROVED** [review]

**Pull request merged**